### PR TITLE
 (non)Differentiable CV parameter-shift gradient transform

### PR DESCRIPTION
**Pull request by josh146**
**Context:** As part of the roadmap for supporting differentiable batch execution, gradient logic will be moved out of the subclasses and into a module of pure functions. This is the third PR after #1476 and #1479; here, we move the CV parameter-shift logic out of `CVParamShiftTape` and into a new gradients package.

**Description of the Change:**

- Adds a function `second_order_param_shift` for computing the second-order CV parameter-shift gradient of a tape terminating in expectation values. Port of `CVQubitParamShift.parameter_shift_second_order`.

- Adds a function `var_param_shift` for computing the CV parameter-shift gradient of a tape terminating in one or more variances. Directly equivalent to `CVQubitParamShift.parameter_shift_var`.

- Adds a wrapper function `param_shift_cv`, directly analagous to `CVQubitParamShift.parameter_shift`, which does the following:

  - Basic input validation
  - Performs static analysis of the tape parameters, to determine which support parameter-shift rules.
  - A fallback mode; `finite_diff` is called for any unsupported parameters
  - Dispatches to one of `expval_param_shift`, `var_param_shift`, or `second_order_param_shift` depending on the structure of the tape
  - Provides a processing function to combine the gradients computed via fallback and the parameter-shift methods.

**Benefits:**

- The parameter-shift logic is now much more user and dev accessible

**Possible Drawbacks:**

- See drawbacks in #1476 
- This transforms is _not_ differentiable, as it depends on a lot of legacy CV code (such as Heisenberg transform functions) which are written in vanilla NumPy and not differentiable.
- The size of the PR - there is very little new code here, it is all a port of `CVParamShiftTape`, including the tests!

**Related GitHub Issues:**

### CODE REVIEW

**Comment by josh146** [comment]
[ch7888]

File: pennylane/gradients/parameter_shift_cv.py [hunk 1]
```diff
@@ -0,0 +1,678 @@
...
+
+import numpy as np
+
+import pennylane as qml
+
+from .finite_difference import finite_diff, generate_shifted_tapes
+from .parameter_shift import expval_param_shift, _get_operation_recipe, _process_gradient_recipe
+
+
+def _grad_method(tape, idx):
```
Comment (josh146) [inline pennylane/gradients/parameter_shift_cv.py:30]: This method is the same as `CVParamShiftTape._grad_method`

File: pennylane/gradients/parameter_shift_cv.py [hunk 2]
```diff
@@ -0,0 +1,678 @@
...
+        # gradient analysis has already been performed on this tape
+        return
+
+    tape._gradient_fn = param_shift_cv
+
+    for idx, info in tape._par_info.items():
+        info["grad_method"] = _grad_method(tape, idx)
+
+
+def _transform_observable(obs, Z, device_wires):
```
Comment (josh146) [inline pennylane/gradients/parameter_shift_cv.py:129]: This method is the same as `CVParamShiftTape._transform_observable`

File: pennylane/gradients/parameter_shift_cv.py [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+    if A.ndim == 2:
+        A = A + A.T
+
+    # TODO: if the A matrix corresponds to a known observable in PennyLane,
+    # for example qml.X, qml.P, qml.NumberOperator, we should return that
+    # instead. This will allow for greater device compatibility.
+    return qml.PolyXP(A, wires=device_wires)
+
+
+def var_param_shift(tape, dev_wires, argnum=None, shift=np.pi / 2, gradient_recipes=None, f0=None):
```
Comment (josh146) [inline pennylane/gradients/parameter_shift_cv.py:166]: This method is the same as `CVParamShiftTape.parameter_shift_var`

File: pennylane/gradients/parameter_shift_cv.py [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+        pdA2 = pdA2_fn(results[tape_boundary:])
+
+        # return d(var(A))/dp = d<A^2>/dp -2 * <A> * d<A>/dp for the variances (mask==True)
+        # d<A>/dp for plain expectations (mask==False)
+        return qml.math.where(mask, pdA2 - 2 * f0 * pdA, pdA)
+
+    return gradient_tapes, processing_fn
+
+
+def second_order_param_shift(tape, dev_wires, argnum=None, shift=np.pi / 2, gradient_recipes=None):
```
Comment (josh146) [inline pennylane/gradients/parameter_shift_cv.py:268]: This method is ported from `CVParamShiftTape.parameter_shift_second_order`, with some modifications (listed below).

File: pennylane/gradients/parameter_shift_cv.py [hunk 5]
```diff
@@ -0,0 +1,678 @@
...
+
+        if not any(i is None for i in constants):
+            # Check if *all* transformed observables corresponds to a constant term.
+            # term. If this is the case for all transformed observables on the tape,
+            # then <psi|A|psi> = A<psi|psi> = A,
+            # and we can avoid the device execution.
+            shapes[-1] = 0
+            obs_indices.append(transformed_obs_idx)
+            gradient_values.append(constants)
+            continue
```
Comment (josh146) [inline pennylane/gradients/parameter_shift_cv.py:405]: This part is new, and corresponds to this TODO: https://github.com/PennyLaneAI/pennylane/blob/ab71001bc55c2945a0945f42131c45d77da3512e/pennylane/tape/cv_param_shift.py#L344

File: pennylane/gradients/parameter_shift_cv.py [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+            g = qml.math.convert_like(g, results[0])
+            if hasattr(g, "dtype") and g.dtype is np.dtype("object"):
+                grads[i] = qml.math.hstack(g)
+
+        return qml.math.T(qml.math.stack(grads))
+
+    return gradient_tapes, processing_fn
+
+
+def param_shift_cv(
```
Comment (josh146) [inline pennylane/gradients/parameter_shift_cv.py:460]: This wrapper/dispatch function is new!

File: tests/gradients/test_parameter_shift_cv.py [hunk 7]
```diff
@@ -0,0 +1,1105 @@
...
+    _grad_method,
+    _gradient_analysis,
+    _transform_observable,
+)
+
+
+hbar = 2
+
+
+class TestGradAnalysis:
```
Comment (josh146) [inline tests/gradients/test_parameter_shift_cv.py:30]: This is a copy and paste of existing tests: https://github.com/PennyLaneAI/pennylane/blob/ab71001bc55c2945a0945f42131c45d77da3512e/tests/tape/test_cv_param_shift.py#L25

File: tests/gradients/test_parameter_shift_cv.py [hunk 8]
```diff
@@ -0,0 +1,1105 @@
...
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Rotation(1.0, wires=0)
+            qml.expval(qml.X(0))
+
+        with pytest.raises(ValueError, match="unknown gradient method"):
+            _grad_method(tape, 0)
+
+
+class TestTransformObservable:
```
Comment (josh146) [inline tests/gradients/test_parameter_shift_cv.py:192]: Copy and paste of existing tests: https://github.com/PennyLaneAI/pennylane/blob/ab71001bc55c2945a0945f42131c45d77da3512e/tests/tape/test_cv_param_shift.py#L195

File: tests/gradients/test_parameter_shift_cv.py [hunk 9]
```diff
@@ -0,0 +1,1105 @@
...
+        # I, X0, Xa, X2, P0, Pa, P2.
+        A = np.diag([-0.5, 0.25, 0.25, 0, 0, 0, 0])
+        expected = A @ Z + Z @ A
+
+        assert isinstance(res, qml.PolyXP)
+        assert res.wires == wires
+        assert np.allclose(res.data[0], expected, atol=tol, rtol=0)
+
+
+class TestParameterShiftLogic:
```
Comment (josh146) [inline tests/gradients/test_parameter_shift_cv.py:271]: Copy and paste of https://github.com/PennyLaneAI/pennylane/blob/ab71001bc55c2945a0945f42131c45d77da3512e/tests/tape/test_cv_param_shift.py#L274

File: tests/gradients/test_parameter_shift_cv.py [hunk 10]
```diff
@@ -0,0 +1,1105 @@
...
+
+        tape.trainable_params = {0, 2}
+        tapes, fn = qml.gradients.param_shift_cv(tape, dev)
+        assert len(tapes) == 0
+
+        grad = fn(dev.batch_execute(tapes))
+        assert np.allclose(grad, [0, 0])
+
+
+class TestExpectationQuantumGradients:
```
Comment (josh146) [inline tests/gradients/test_parameter_shift_cv.py:391]: Copy and paste of: https://github.com/PennyLaneAI/pennylane/blob/ab71001bc55c2945a0945f42131c45d77da3512e/tests/tape/test_cv_param_shift.py#L356

File: tests/gradients/test_parameter_shift_cv.py [hunk 11]
```diff
@@ -0,0 +1,1105 @@
...
+
+        assert tape._par_info[0]["grad_method"] == "A"
+        assert tape._par_info[1]["grad_method"] == "A"
+
+        # the different methods agree
+        assert np.allclose(grad_A, grad_F, atol=tol, rtol=0)
+        assert np.allclose(grad_A2, grad_F, atol=tol, rtol=0)
+
+
+class TestVarianceQuantumGradients:
```
Comment (josh146) [inline tests/gradients/test_parameter_shift_cv.py:737]: Copy and paste of https://github.com/PennyLaneAI/pennylane/blob/ab71001bc55c2945a0945f42131c45d77da3512e/tests/tape/test_cv_param_shift.py#L702

File: tests/gradients/test_parameter_shift_cv.py [hunk 12]
```diff
@@ -0,0 +1,1105 @@
...
+            qml.var(qml.TensorN(wires=[0]))
+
+        tape.trainable_params = {0, 1}
+        tapes, fn = param_shift_cv(tape, dev)
+        grad = fn(dev.batch_execute(tapes))
+        expected = np.array([2 * a ** 2 + 2 * n + 1, 2 * a * (2 * n + 1)])
+        assert np.allclose(grad, expected, atol=tol, rtol=0)
+
+
+class TestParamShiftInterfaces:
```
Comment (josh146) [inline tests/gradients/test_parameter_shift_cv.py:961]: This is new!!

File: pennylane/gradients/finite_difference.py [hunk 13]
```diff
@@ -161,7 +161,7 @@ def generate_shifted_tapes(tape, idx, shifts, multipliers=None):
         ``idx`` shifted by consecutive values of ``shift``. The length
         of the returned list of tapes will match the length of ``shifts``.
     """
-    params = tape.get_parameters()
+    params = list(tape.get_parameters())
```
Comment (antalszava) [inline pennylane/gradients/finite_difference.py:164]: How come this was changed?

File: pennylane/gradients/parameter_shift_cv.py [hunk 14]
```diff
@@ -0,0 +1,678 @@
...
+
+        for obs in observable_descendents:
+            # get the index of the descendent observable
+            idx = tape.observables.index(obs)
+            transformed_obs_idx.append(idx)
+
+            transformed_obs = _transform_observable(obs, Z, dev_wires)
+
+            A = transformed_obs.parameters[0]
+            constants.append(None)
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:383]: (Minor): how about appending `None` only when there is no constant?

If I understand it correctly, first we append `None` and then remove it in cases when there is a constant term. This logic seems to be slightly more prone to confusion.

File: pennylane/gradients/parameter_shift_cv.py [hunk 15]
```diff
@@ -0,0 +1,678 @@
...
+    def processing_fn(results):
+        grads = []
+        start = 0
+
+        if not results:
+            results = [np.zeros([tape.output_dim])]
+
+        interface = qml.math.get_interface(results[0])
+
+        for i, (s, k, ind) in enumerate(zip(shapes, gradient_values, obs_indices)):
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:420]: Renaming `i, (s, k, ind)` could help with reading this part of the code. If this line becomes too long when renaming, `enumerate(zip(shapes, gradient_values, obs_indices))` could be assigned a name in the previous line.

File: pennylane/gradients/parameter_shift_cv.py [hunk 16]
```diff
@@ -0,0 +1,678 @@
...
+                g = qml.math.zeros_like(results[0], like=interface)
+
+                if k:
+                    g = qml.math.scatter_element_add(g, ind, k, like=interface)
+
+                grads.append(g)
+                continue
+
+            res = results[start : start + s]
+            start = start + s
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:433]: Previously in this PR, there seems to have been a bug due to confusion between `res` and `results` in the finite diff logic. How about renaming `res` to `res_for_observable` or so?

File: pennylane/gradients/parameter_shift_cv.py [hunk 17]
```diff
@@ -0,0 +1,678 @@
...
+
+            # compute the linear combination of results and coefficients
+            res = qml.math.stack(res[0])
+            g = qml.math.zeros_like(res, like=interface)
+
+            if qml.math.get_interface(g) not in ("tensorflow", "autograd"):
+                ind = (ind,)
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:440]: Why is this required?

File: pennylane/gradients/parameter_shift_cv.py [hunk 18]
```diff
@@ -0,0 +1,678 @@
...
+
+            If ``None``, the default gradient recipe containing the two terms
+            :math:`[c_0, a_0, s_0]=[1/2, 1, \pi/2]` and :math:`[c_1, a_1,
+            s_1]=[-1/2, 1, -\pi/2]` is assumed for every parameter.
+        fallback_fn (None or Callable): a fallback grdient function to use for
+            any parameters that do not support the parameter-shift rule.
+        f0 (tensor_like[float] or None): Output of the evaluated input tape. If provided,
+            and the gradient recipe contains an unshifted term, this value is used,
+            saving a quantum evaluation.
+        force_order2 (bool): iff True, use the order-2 method even if not necessary
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:503]:
```suggestion
        force_order2 (bool): if True, use the order-2 method even if not necessary
```

File: pennylane/gradients/parameter_shift_cv.py [hunk 19]
```diff
@@ -0,0 +1,678 @@
...
+
+    This transform supports analytic gradients of Gaussian CV operations using
+    the parameter-shift rule. This gradient method returns *exact* gradients,
+    and can be computed directly on quantum hardware.
+
+    Analytic gradients of photonic circuits that satisfy
+    the following constraints with regards to measurements are supported:
+
+    * Expectation values are restricted to observables that are first- and
+      second-order in :math:`\hat{x}` :math:`\hat{p}` only.
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:518]:
```suggestion
      second-order in :math:`\hat{x}` and :math:`\hat{p}` only.
```

File: pennylane/gradients/parameter_shift_cv.py [hunk 20]
```diff
@@ -0,0 +1,678 @@
...
+      This includes :class:`~.X`, :class:`~.P`, :class:`~.QuadOperator`,
+      :class:`~.PolyXP`, and :class:`~.NumberOperator`.
+
+      For second-order observables, the device **must support** :class:`~.PolyXP`.
+
+    * Variances are restricted to observables that are first-order
+      in :math:`\hat{x}` :math:`\hat{p}` only. This includes :class:`~.X`, :class:`~.P`,
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:525]:
```suggestion
      in :math:`\hat{x}` and :math:`\hat{p}` only. This includes :class:`~.X`, :class:`~.P`,
```

File: pennylane/gradients/parameter_shift_cv.py [hunk 21]
```diff
@@ -0,0 +1,678 @@
...
+        return gradient_tapes, lambda _: np.zeros([tape.output_dim, len(tape.trainable_params)])
+
+    # TODO: replace the JacobianTape._choose_params_with_methods
+    # functionality before deprecation.
+    method_map = dict(tape._choose_params_with_methods(diff_methods, argnum))
+    var_present = any(m.return_type is qml.operation.Variance for m in tape.measurements)
+
+    unsupported_params = [idx for idx, g in method_map.items() if g == "F"]
+    first_order_params = [idx for idx, g in method_map.items() if g == "A"]
+    second_order_params = [idx for idx, g in method_map.items() if g == "A2"]
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:619]: (Minor): guess it's not super important, though would having a single for loop be more performant? We seem to be iterating through the same map 3 times.

File: pennylane/gradients/parameter_shift_cv.py [hunk 22]
```diff
@@ -0,0 +1,678 @@
...
+        return gradient_tapes, fns[-1]
+
+    gradient_recipes = gradient_recipes or [None] * len(argnum)
+
+    if var_present:
+        _update(var_param_shift(tape, dev.wires, argnum, shift, gradient_recipes, f0))
+
+    else:
+        if first_order_params:
+            _update(expval_param_shift(tape, first_order_params, shift, gradient_recipes, f0))
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:659]: How come `dev.wires` is not required for `expval_param_shift`?

File: pennylane/gradients/parameter_shift_cv.py [hunk 23]
```diff
@@ -0,0 +1,678 @@
...
+    if not argnum:
+        # No analytic parameters. Return the existing fallback tapes/fn
+        return gradient_tapes, fns[-1]
+
+    gradient_recipes = gradient_recipes or [None] * len(argnum)
+
+    if var_present:
+        _update(var_param_shift(tape, dev.wires, argnum, shift, gradient_recipes, f0))
+
+    else:
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift_cv.py:657]:
```suggestion
    else:
        # Only expectation values were specified
```

File: tests/tape/test_jacobian_tape.py [hunk 24]
```diff
@@ -456,6 +456,7 @@ def test_choose_params_and_methods(self, diff_methods, argnum):
 
         assert len(res) == num_params
 
+    @pytest.mark.xfail
```
Comment (antalszava) [inline tests/tape/test_jacobian_tape.py:459]: Should this test be deleted if we've decided to remove the check altogether?

**Review by antalszava — COMMENTED** [review]
@josh146 awesome :slightly_smiling_face:  :muscle: The chat and the comments in the PR were very useful. Nothing major from my side, might just check back to check your thoughts on the comments posted.

File: pennylane/gradients/finite_difference.py (hunk 13 above)
Comment (josh146) [reply pennylane/gradients/finite_difference.py:164]: Oh, I discovered during this PR that sometimes `tape.get_parameters()` will return an Autograd `SequenceBox`, for some reason. I was unable to work out what causes this to happen, but making sure that params is always cast to a list fixes it (for now).

I might be able to better track it down once the new interfaces are finished

File: pennylane/gradients/parameter_shift_cv.py (hunk 14 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:383]: I originally did that, but then it causes issues further down as you are unable to use the `constants` list in the `zip` 🤔 

So it seems to be a tradeoff; make it simpler here at the expense of more complicated `zip` logic later, or make it longer here so that it works in the `zip`.

File: pennylane/gradients/parameter_shift_cv.py (hunk 21 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:619]: Good point!

File: pennylane/gradients/parameter_shift_cv.py (hunk 22 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:659]: It's a unique requirement for the CV rule, since the transformation of the observable requires the device wires

File: tests/tape/test_jacobian_tape.py (hunk 24 above)
Comment (josh146) [reply tests/tape/test_jacobian_tape.py:459]: Good idea!

File: pennylane/gradients/parameter_shift_cv.py (hunk 15 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:420]: Good idea!

Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:420]: Done :)

File: pennylane/gradients/parameter_shift_cv.py (hunk 16 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:433]: Good idea, done!

File: pennylane/gradients/parameter_shift_cv.py (hunk 17 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:440]: I'm not 100% sure - as far as I can tell, the scattering function in TF and Autograd requires an additional dimension, whereas Torch and JAX don't. I couldn't find a combination that work for all :(

File: pennylane/gradients/parameter_shift_cv.py (hunk 21 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:619]: Have updated this to a single for loop :)

File: pennylane/gradients/parameter_shift_cv.py (hunk 22 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:659]: It is very annoying, I must say :( This is inherited from `CVParamShiftTape`

File: tests/tape/test_jacobian_tape.py (hunk 24 above)
Comment (josh146) [reply tests/tape/test_jacobian_tape.py:459]: Done :)

File: pennylane/gradients/parameter_shift_cv.py (hunk 14 above)
Comment (antalszava) [reply pennylane/gradients/parameter_shift_cv.py:383]: Oh really? I would have expected that modifying the if structure to always append to `constants` and extending it with two `else` branches could help :thinking: 
```python
            # Check if the transformed observable corresponds to a constant term.
            if len(A.nonzero()[0]) == 1:
                if A.ndim == 2 and A[0, 0] != 0:
                    constants.append(A[0, 0])

                elif A.ndim == 1 and A[0] != 0:
                    constants.append(A[0])
                else:
                    constants.append(None)
            else:
                constants.append(None)
```

**Review by antalszava — APPROVED** [review]
Looks good to me! :slightly_smiling_face:  :100: :tada:

Just came to mind: should we note in the docs that differentiation is not always supported for the CV gradient transform?

File: pennylane/gradients/parameter_shift_cv.py (hunk 14 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:383]: Ah, I see what you mean @antalszava!

I wonder if it is slightly slower though, due to the cost of the additional branching? Not sure, and if it is, it is probably very small 😆

Comment (antalszava) [reply pennylane/gradients/parameter_shift_cv.py:383]: > I wonder if it is slightly slower though, due to the cost of the additional branching?

Good question! Maybe since these are `else`s, those are less costly. With the current version, we might do some extra list operations. Not sure about the overall performance implications. :thinking:

File: pennylane/gradients/parameter_shift_cv.py [hunk 25]
```diff
@@ -0,0 +1,690 @@
...
+        return "0"
+
+    if "F" in best:
+        # one non-analytic observable path makes the whole operation
+        # gradient method fallback to finite-difference
+        return "F"
+
+    if "A2" in best:
+        # one second order observable makes the whole operation gradient
+        # require the second order parameter-shift rule
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift_cv.py:109]: I think "second order" -> "second-order" throughout (same with first)?

File: pennylane/gradients/parameter_shift_cv.py [hunk 26]
```diff
@@ -0,0 +1,690 @@
...
+        return
+
+    tape._gradient_fn = param_shift_cv
+
+    for idx, info in tape._par_info.items():
+        info["grad_method"] = _grad_method(tape, idx)
+
+
+def _transform_observable(obs, Z, device_wires):
+    """Apply a Gaussian linear transformation to each index of an observable.
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift_cv.py:130]: The transformation is applied to the observable itself, right, not the index?

File: pennylane/gradients/parameter_shift_cv.py [hunk 27]
```diff
@@ -0,0 +1,690 @@
...
+
+      For second-order observables, the device **must support** :class:`~.PolyXP`.
+
+    * Variances are restricted to observables that are first-order
+      in :math:`\hat{x}` and :math:`\hat{p}` only. This includes :class:`~.X`, :class:`~.P`,
+      :class:`~.QuadOperator`, and *some* parameter values of :class:`~.PolyXP`.
+
+      The device **must support** :class:`~.PolyXP`.
+
+    Fock state probabilities (tapes that return :func:`~pennylane.probs` or
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift_cv.py:531]: Maybe worth putting this in a little warning box for emphasis.

File: tests/gradients/test_parameter_shift_cv.py [hunk 28]
```diff
@@ -0,0 +1,1105 @@
...
+        """Tests that the output of the parameter-shift CV transform
+        can be differentiated using JAX, yielding second derivatives."""
+        jax = pytest.importorskip("jax")
+        from jax import numpy as jnp
+        from pennylane.interfaces.jax import JAXInterface
+        from jax.config import config
+
+        config.update("jax_enable_x64", True)
+
+        dev = qml.device("default.gaussian", wires=2)
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift_cv.py:1069]: I noticed this device has 2 wires but for the other interfaces there is only 1. Should we test with different numbers of wires and/or different sets of operations in the circuits over the different interface? (and maybe the different cases for the gradients, like the cases where everything falls back to finite diff vs. when analytic gradients are supported?)

**Review by glassnotes — COMMENTED** [review]
Just leaving some comments after a first pass, thanks for marking the stuff that is new! It all looks good, I will just do a second pass tomorrow, moreso for my own understanding :+1:

File: tests/tape/test_jacobian_tape.py [hunk 29]
```diff
@@ -456,19 +456,6 @@ def test_choose_params_and_methods(self, diff_methods, argnum):
 
         assert len(res) == num_params
 
-    @pytest.mark.parametrize("argnum", [1, 2, 3, -1])
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/tape/test_jacobian_tape.py:459]: This test corresponded to the incorrect validation that was removed from `jacobian_tape` above :)

image

File: pennylane/gradients/parameter_shift_cv.py (hunk 26 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:130]: Yes I think you're correct
```suggestion
    """Apply a Gaussian linear transformation to an observable.
```

File: pennylane/gradients/parameter_shift_cv.py (hunk 25 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:109]: Oops, will update!

File: pennylane/gradients/parameter_shift_cv.py (hunk 14 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:383]: I updated it to this, which I think is cleaner :)

image

File: pennylane/gradients/parameter_shift_cv.py (hunk 27 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:531]: done!

File: pennylane/gradients/parameter_shift_cv.py [hunk 30]
```diff
@@ -0,0 +1,690 @@
...
+    expectation values of :class:`~.FockStateProjector`) are not supported.
```
Comment (josh146) [inline pennylane/gradients/parameter_shift_cv.py:532]:
```suggestion
    .. warning::

        Fock state probabilities (tapes that return :func:`~pennylane.probs` or
        expectation values of :class:`~.FockStateProjector`) are not supported.
```

File: pennylane/gradients/parameter_shift_cv.py [hunk 31]
```diff
@@ -0,0 +1,694 @@
...
+    dev,
+    argnum=None,
+    shift=np.pi / 2,
+    gradient_recipes=None,
+    fallback_fn=finite_diff,
+    f0=None,
+    force_order2=False,
+):
+    r"""Generate the CV parameter-shift tapes and postprocessing methods required
+    to compute the gradient of an gate parameter with respect to the CV
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift_cv.py:474]:
```suggestion
    to compute the gradient of a gate parameter with respect to the CV

```

File: pennylane/gradients/parameter_shift_cv.py [hunk 32]
```diff
@@ -0,0 +1,694 @@
...
+    output.
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift_cv.py:475]:
```suggestion
    to compute the gradient of a gate parameter with respect to the CV output.
```

File: tests/gradients/test_parameter_shift_cv.py [hunk 33]
```diff
@@ -0,0 +1,1105 @@
...
+        assert _grad_method(tape, 2) == "F"
+        assert _grad_method(tape, 3) == "F"
+
+    def test_second_order_expectation(self):
+        """Test that the expectation of a second-order observable forces
+        the gradient method to use the second-order parameter-shift rule"""
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Rotation(1.0, wires=[0])
+            qml.expval(qml.NumberOperator(0))  # second order
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift_cv.py:174]: In line 160 above, `var(qml.NumberOperator(0))` is marked as second-order, but here `expval` is, should this be `var` instead?

File: tests/gradients/test_parameter_shift_cv.py [hunk 34]
```diff
@@ -0,0 +1,1105 @@
...
+        This ensures that, assuming their _heisenberg_rep is defined, the quantum
+        gradient analytic method can still be used, and returns the correct result.
+
+        Currently, the only such operation is qml.Interferometer. In the future,
+        we may consider adding a qml.GaussianTransfom operator.
+        """
+
+        if t == 1:
+            pytest.xfail(
+                "There is a bug in the second order CV parameter-shift rule; "
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift_cv.py:679]: Just noticed this section, is this bug still present? :cry:

File: tests/gradients/test_parameter_shift_cv.py [hunk 35]
```diff
@@ -0,0 +1,1105 @@
...
+
+        tape.trainable_params = {0, 1}
+        tapes, fn = param_shift_cv(tape, dev)
+        grad = fn(dev.batch_execute(tapes))
+        expected = np.array([2 * a ** 2 + 2 * n + 1, 2 * a * (2 * n + 1)])
+        assert np.allclose(grad, expected, atol=tol, rtol=0)
+
+
+class TestParamShiftInterfaces:
+    """Test that the transform is differentiable"""
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift_cv.py:962]: I know I asked the other day, but could you please remind me the purpose of the differentiation tests and how they work when the PR states explicitly that the transform is not differentiable?

**Review by glassnotes — APPROVED** [review]
Have left a couple questions, but otherwise this looks good! :+1:

File: tests/gradients/test_parameter_shift_cv.py (hunk 33 above)
Comment (josh146) [reply tests/gradients/test_parameter_shift_cv.py:174]: Oops - this one is correct since `<n> = <X^2 + P^2 - hbar>` is second order in X and P. However, `Var(n) = <n^2> - <n>^2` is thus 4th order and the comment on line 160 is incorrect

File: tests/gradients/test_parameter_shift_cv.py [hunk 36]
```diff
@@ -0,0 +1,1105 @@
...
+            qml.Rotation(1.0, wires=[0])
+            qml.var(qml.NumberOperator(0))  # second order
+
+        assert _grad_method(tape, 0) == "F"
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Rotation(1.0, wires=[0])
+            qml.Rotation(1.0, wires=[1])
+            qml.Beamsplitter(0.5, 0.0, wires=[0, 1])
+            qml.var(qml.NumberOperator(0))  # second order
```
Comment (josh146) [inline tests/gradients/test_parameter_shift_cv.py:160]:
```suggestion
            qml.var(qml.NumberOperator(0))  # fourth order
```

File: tests/gradients/test_parameter_shift_cv.py (hunk 34 above)
Comment (josh146) [reply tests/gradients/test_parameter_shift_cv.py:679]: Yes, unfortunately :(

File: tests/gradients/test_parameter_shift_cv.py (hunk 35 above)
Comment (josh146) [reply tests/gradients/test_parameter_shift_cv.py:962]: Ah! So. All my intuition says that the transform should not be differentiable, because so much of it uses vanilla numpy.

I wanted to write these tests anyway, and just mark them `xfail`, so we had tests ready in case we needed them in the future (and to remind us that this doesn't work).

But, the tests.... worked????

Against all my intuition.

I don't really trust them though, I am sure there are many edge cases where autodiff would fail... I just can't find them haha
